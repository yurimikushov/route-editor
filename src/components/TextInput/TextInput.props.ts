type TextInputProps = {
  className?: string
  type?: 'text' | 'search' | 'email' | 'tel'
  value: string
  placeholder?: string
  hasClear?: boolean
  onChange: (value: string) => void
  onFocus?: () => void
  onBlur?: () => void
}

export default TextInputProps
