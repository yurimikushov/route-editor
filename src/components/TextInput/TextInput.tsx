import {
  ChangeEvent,
  ForwardRefRenderFunction,
  forwardRef,
  useRef,
  useImperativeHandle,
} from 'react'
import cn from 'classnames'
import isEmpty from 'lodash/isEmpty'
import nonNullable from 'lib/nonNullable'
import useEffectWhen from 'hooks/useEffectWhen'
import ClearButton from 'components/ClearButton'
import TextInputProps from './TextInput.props'

const TextInput: ForwardRefRenderFunction<HTMLInputElement, TextInputProps> = (
  {
    className,
    type = 'text',
    value,
    autoFocus = false,
    hasClear = false,
    onChange,
    ...props
  },
  externalRef
) => {
  const inputRef = useRef<HTMLInputElement>(null)

  useImperativeHandle(externalRef, () => nonNullable(inputRef.current), [])

  useEffectWhen(() => {
    inputRef.current?.focus()
  }, autoFocus)

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    onChange(e.target.value)
  }

  const handleClear = () => {
    inputRef.current?.focus()
    onChange('')
  }

  const shouldDisplayClear = hasClear && !isEmpty(value)

  return (
    <div
      className={cn(
        className,
        'px-3 py-2',
        'inline-flex items-center',
        'border-2 border-gray-300 rounded-md',
        'focus-within:ring-2 focus-within:ring-indigo-500 focus-within:border-indigo-500'
      )}
    >
      <input
        {...props}
        ref={inputRef}
        type={type}
        className={cn(className, 'focus:outline-none', {
          'w-[calc(100%-theme(space.5))]': shouldDisplayClear,
        })}
        value={value}
        onChange={handleChange}
      />
      {shouldDisplayClear && (
        <ClearButton
          className='w-5 h-5 min-w-5 min-h-5'
          onClick={handleClear}
        />
      )}
    </div>
  )
}

export default forwardRef(TextInput)
