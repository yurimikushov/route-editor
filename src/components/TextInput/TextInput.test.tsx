import { fireEvent, render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import noop from 'lodash/noop'
import TextInput from './TextInput'

it('should render text input', () => {
  const tree = renderer.create(<TextInput value='' onChange={noop} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should render text input without clear btn', () => {
  const tree = renderer
    .create(
      <TextInput
        value='123'
        placeholder='Type here'
        onChange={noop}
        onFocus={noop}
        onBlur={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should render text input with clear btn', () => {
  const tree = renderer
    .create(<TextInput value='123' hasClear onChange={noop} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should call onChange when some text is typed', () => {
  let currentValue = ''
  const handleChange = jest.fn((value: string) => {
    currentValue = value
  })

  const component = render(
    <TextInput
      value={currentValue}
      placeholder='Type here'
      onChange={handleChange}
    />
  )

  const textInput = screen.getByDisplayValue(currentValue)

  fireEvent.change(textInput, { target: { value: 'q' } })
  fireEvent.change(textInput, { target: { value: 'qw' } })
  fireEvent.change(textInput, { target: { value: 'qwe' } })
  fireEvent.change(textInput, { target: { value: 'qwer' } })
  fireEvent.change(textInput, { target: { value: 'qwert' } })
  fireEvent.change(textInput, { target: { value: 'qwerty' } })

  expect(currentValue).toBe('qwerty')
  expect(handleChange).toHaveBeenCalledTimes(6)

  currentValue = '123456'
  component.rerender(<TextInput value={currentValue} onChange={handleChange} />)

  fireEvent.change(textInput, { target: { value: '12345' } })
  fireEvent.change(textInput, { target: { value: '1234' } })

  expect(currentValue).toBe('1234')
  expect(handleChange).toHaveBeenCalledTimes(8)
})

it('should call onChange with empty value when clear btn is clicked', () => {
  let currentValue = 'qwerty'
  const handleChange = jest.fn((value: string) => {
    currentValue = value
  })

  render(
    <TextInput
      value={currentValue}
      placeholder='Type here'
      hasClear
      onChange={handleChange}
    />
  )

  fireEvent.click(screen.getByTestId('clear-text-input'))

  expect(currentValue).toBe('')
  expect(handleChange).toHaveBeenCalledTimes(1)
})

it('should focus input when clear btn is clicked', () => {
  render(
    <TextInput
      value='qwerty'
      placeholder='Type here'
      hasClear
      onChange={noop}
    />
  )

  fireEvent.click(screen.getByTestId('clear-text-input'))

  expect(document.activeElement).toBe(screen.getByDisplayValue('qwerty'))
})
