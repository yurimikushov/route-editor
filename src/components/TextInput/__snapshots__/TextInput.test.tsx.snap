// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render text input 1`] = `
<div
  className="px-3 py-2 inline-flex items-center border-2 border-gray-300 rounded-md focus-within:ring-2 focus-within:ring-indigo-500 focus-within:border-indigo-500"
>
  <input
    className="focus:outline-none"
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;

exports[`should render text input with clear btn 1`] = `
<div
  className="px-3 py-2 inline-flex items-center border-2 border-gray-300 rounded-md focus-within:ring-2 focus-within:ring-indigo-500 focus-within:border-indigo-500"
>
  <input
    className="focus:outline-none w-[calc(100%-theme(space.5))]"
    onChange={[Function]}
    type="text"
    value="123"
  />
  <button
    className="w-5 h-5 min-w-5 min-h-5"
    data-testid="clear-text-input"
    onClick={[Function]}
  >
    Clear
  </button>
</div>
`;

exports[`should render text input without clear btn 1`] = `
<div
  className="px-3 py-2 inline-flex items-center border-2 border-gray-300 rounded-md focus-within:ring-2 focus-within:ring-indigo-500 focus-within:border-indigo-500"
>
  <input
    className="focus:outline-none"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type here"
    type="text"
    value="123"
  />
</div>
`;
